### PR TITLE
feat: add e2e testing framework

### PR DESCRIPTION
**Pull request by marcoppenheimer**
## TODO
- Update the README with the info in this far-too-long PR description.
- Also not sure why the CI isn't running, ideas welcome.

##### `feat: add e2e testing framework`
- see dedicated section for more detail for more detail
##### `refactor: remove existing bundle tests`
- After https://github.com/canonical/zookeeper-operator/pull/47, the bundle is much more stable and no longer needs dedicated 'bundle' tests. They have been kept in for posterity but could be removed
- They have been superseded by the (upcoming) e2e testing 
##### `style: use ruff and poetry`
- This is just to keep the repo in line with the Kafka/ZooKeeper repos
 

# E2E Testing Framework

### Running the tests
To run tests with different configurations, you can pass arguments to the `tox -e interation-e2e -- ` command. These arguments are parsed by `tests/integration/e2e/conftest.py` as `scope=module` arguments, defaulting to what is set in `tests/integration/e2e/literals.py` if not set

Optional flags for fresh deployments are as follows:
- `--tls` for running the cluster with SSL encryption

For example: 
```bash
# normal:
tox -e integration-e2e

# with tls:
tox -e integration-e2e -- --tls
```

### Pre-existing deployments
Essential flags for running tests on a pre-existing deployment (e.g for validation) are as follows:
- `--kafka` and `--zookeeper` with values of their pre-existing Juju Application names
- `--model` with value of the pre-existing Juju model they're deployed under
- `--no-deploy` to skip deployment of a new cluster, and teardown at test completion

For example:
```bash
# normal:
tox -e integration-e2e -- --kafka=<KAFKA_APP_NAME> --zookeeper=<ZOOKEEPER_APP_NAME> --model=<JUJU_MODEL_NAME> --no-deploy

# with tls:
tox -e integration-e2e -- --tls --kafka=<KAFKA_APP_NAME> --zookeeper=<ZOOKEEPER_APP_NAME> --model=<JUJU_MODEL_NAME> --no-deploy
```
A commented out example of this is in `tox:integration-e2e-existing`, purely for demonstration purposes. Will be updated once the tests are more fleshed out.

### Why do I have to pass Kafka + ZooKeeper flags?
Pre-existing deployments might not have applications called `kafka` or `zookeeper` or `tls-certificates-operator`. By passing the name, it gets loaded as a Pytest fixture name that we can rely on during relations with, say, a `producer` client. If not set, they default to the main Charm name (except `tls-certificates-operator`, which defaults to `certificates` because the former is long). 

In order to use these, you can do something similar to the following:
```python
async def test_cluster_is_deployed_successfully(
    ops_test: OpsTest, kafka, zookeeper, tls, certificates
):
    assert ops_test.model.applications[kafka].status == "active"
    assert ops_test.model.applications[zookeeper].status == "active"

    if tls:
        assert ops_test.model.applications[certificates].status == "active"
```

## Writing tests using factory fixtures
Fixtures can be found in `tests/integration/e2e/conftest.py` for `deploy_cluster` and `deploy_client`. All `pytest` fixtures evaluate before test-start, caching the returned values to be shared across their defined `scope`. 

### Deploying the Cluster
`deploy_cluster` has `scope="module"`, and when it is first referenced in a module (e.g as a function fixture), it will run to completion, either calling a nested function `_deploy_non_tls_cluster` or `_deploy_tls_cluster` based on the passed `--tls` flag. Each of the main 3 applications (Kafka, ZooKeeper, tls-certificates-operator) will run in their own async operations to hopefully speed up deployment.

Although somewhat obscured, as the fixture doesn't need to be re-referenced, returning the coroutine to be fully executed on test-module set-up within `_deploy_X_cluster` is fine, and gives us the option of extending it if ever we want to deploy across multiple models concurrently. Unlikely however with the state of our CI.

Tests decorated with `@pytest.mark.skip_if_deployed` will not run if `--no-deploy` and `--model` have been passed.

Make sure to start your test modules with something similar to the following:
```python
@pytest.mark.skip_if_deployed
async def test_deploy(ops_test: OpsTest, deploy_cluster):
    await asyncio.sleep(0)  # do nothing, await deploy_cluster
```

### Deploying Producer/Consumer Clients
The `deploy_client` fixture has `scope="function"`, and uses a [Factory pattern](https://docs.pytest.org/en/6.2.x/fixture.html#factories-as-fixtures) to yield a coroutine which can be then awaited from within the test. This allows you to pass `role` as an argument, and pass as many instances of that client as you like.

When executed, a nested function `_deploy_client(role)` generates a UUID and names the application `role-UUID`, e.g `producer-xciv` to avoid app-name clashes Juju side. It also relates to Kafka and waits for Kafka to `active/idle`. It passes back application name (e.g `producer-xciv`) so that you can reference it from within `ops_test` for whatever actions/scaling you might want to do in the actual test.

On completion of the actual test function, all deployed clients from within the function will be `juju remove`'d as clean-up.

To use these clients in your test modules, do something similar to the following:
```python
async def test_clients_actually_set_up(ops_test: OpsTest, deploy_client):
    producer = await deploy_client(role="producer")
    consumer = await deploy_client(role="consumer")

    assert ops_test.model.applications[consumer].status == "active"
    assert ops_test.model.applications[producer].status == "active"


async def test_clients_actually_tear_down_after_test_exit(ops_test: OpsTest):
    assert "consumer" not in "".join(ops_test.model.applications.keys())
    assert "producer" not in "".join(ops_test.model.applications.keys())
```

### CODE REVIEW

File: tests/integration/e2e/fixtures.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        ),
+    )
+    return await ops_test.model.wait_for_idle(
+        apps=["kafka", "zookeeper"], idle_period=10, timeout=300
+    )
+
+
+@pytest.fixture
+async def tls_cluster(ops_test: OpsTest, cluster):
+    await cluster
```
Comment (marcoppenheimer) [inline tests/integration/e2e/fixtures.py:31]: fixtures dependent on eachother

File: tests/integration/e2e/test_e2e.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+import pytest
+from pytest_operator.plugin import OpsTest
+
+logger = logging.getLogger(__name__)
+
+
+@pytest.mark.skip_if_deployed
+async def test_deploy(ops_test: OpsTest, cluster, tls_cluster, tls):
+    if tls:
```
Comment (marcoppenheimer) [inline tests/integration/e2e/test_e2e.py:16]: flag used here

File: tests/integration/e2e/test_e2e.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+import asyncio
+import logging
+
+import pytest
+from pytest_operator.plugin import OpsTest
+
+logger = logging.getLogger(__name__)
+
+
+@pytest.mark.skip_if_deployed
```
Comment (marcoppenheimer) [inline tests/integration/e2e/test_e2e.py:14]: will not run if `--model=` and `--no-deploy` passed on `tox -e integration-e2e`

Comment (deusebio) [reply tests/integration/e2e/test_e2e.py:14]: **praise** I like that we can toggle things using cli !

File: tox.ini [hunk 4]
```diff
@@ -76,5 +76,17 @@ deps =
     pytest-operator
     -r{toxinidir}/requirements.txt
 commands =
-    pytest -v --tb native --log-cli-level=INFO -s {posargs}
+    pytest -k "bundle" -v --tb native --log-cli-level=INFO -s {posargs}
```
Comment (deusebio) [inline tox.ini:79]: **question** what does this "bundle" refer to? I'm not sure I undersdtand its usage

File: tox.ini [hunk 5]
```diff
@@ -76,5 +76,17 @@ deps =
...
+[testenv:integration-e2e]
+description = Run vm e2e integration tests. Add `--tls` to specify tls tests. Add `--model=<MODEL> --no-deploy` to run on existing deployments.
+deps =
+    pytest
+    juju==2.9.38.1
+    kazoo
+    pure-sasl
+    tenacity
+    pytest-operator
+    -r{toxinidir}/requirements.txt
```
Comment (deusebio) [inline tox.ini:90]: maybe using poetry here?

**Review by deusebio — APPROVED** [review]
I really like this structuring and PR. 

The only thing (apart from a rather minor comment) would be that I believe it was really good to put some efforts on writing a good description of the PR. However, it may be worth putting this in the PR itself, like a markdown in `docs/e2e_testing_guide.md`: i can really see myself (and developers in general) going back there from time to time

**Review by welpaolo — APPROVED** [review]
Looks really good to me!

File: tests/integration/e2e/conftest.py [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+        await _deploy_non_tls_cluster()
+
+
+@pytest.fixture(scope="function")
+async def deploy_client(ops_test: OpsTest, kafka):
+    """Factory fixture for deploying + tearing down client applications."""
+    # tracks deployed app names for teardown later
+    apps = []
+
+    async def _deploy_client(role: Literal["producer", "consumer"]):
```
[the remark this answers is not in the document]
Comment (marcoppenheimer) [reply tests/integration/e2e/conftest.py:141]: I'd vote leaving it out of this PR as it's a fairly small change. I think we're going to extend the 'bones' of the skeleton a fair bit anyway already so maybe worth addressing later?

**Comment by deusebio** [comment]
@marcoppenheimer can you rebase the PR and fix the conflicts, such that we can merge this PR?

**Review by deusebio — APPROVED** [review]

**Review by welpaolo — APPROVED** [review]
LGTM

**Pull request merged**